src/runtime/inorder_executor.cpp

```diff
@@ -0,0 +1,222 @@
+ *    list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <cassert>
+
+#include "hipSYCL/runtime/inorder_executor.hpp"
+#include "hipSYCL/runtime/inorder_queue.hpp"
+#include "hipSYCL/runtime/operations.hpp"
+#include "hipSYCL/runtime/serialization/serialization.hpp"
+
+namespace hipsycl {
+namespace rt {
+
+namespace {
+
+class queue_operation_dispatcher : public operation_dispatcher
+{
+public:
+  queue_operation_dispatcher(inorder_queue* q)
+  : _queue{q}
+  {}
+
+  virtual ~queue_operation_dispatcher(){}
+
+  virtual result dispatch_kernel(kernel_operation *op,
+                                 dag_node_ptr node) final override {
+
+    return _queue->submit_kernel(*op, node);
+  }
+
+  virtual result dispatch_memcpy(memcpy_operation *op,
+                                 dag_node_ptr node) final override {
+    return _queue->submit_memcpy(*op, node);
+  }
+
+  virtual result dispatch_prefetch(prefetch_operation *op,
+                                   dag_node_ptr node) final override {
+    return _queue->submit_prefetch(*op, node);
+  }
+
+  virtual result dispatch_memset(memset_operation *op,
+                                 dag_node_ptr node) final override {
+    return _queue->submit_memset(*op, node);
+  }
+
+private:
+  inorder_queue* _queue;
+};
+
+std::size_t
+get_maximum_execution_index_for_lane(const std::vector<dag_node_ptr> &nodes,
+                                     inorder_queue* lane) {
+  std::size_t index = 0;
+  for (const auto &node : nodes) {
+    if (node->is_submitted() &&
+        node->get_assigned_device().get_backend() ==
+            lane->get_device().get_backend() &&
+        node->get_assigned_execution_lane() == lane) {
+      if(node->get_assigned_execution_index() > index)
+        index = node->get_assigned_execution_index();
+    }
+  }
+  return index;
+}
+
+} // anonymous namespace
+
+inorder_executor::inorder_executor(std::unique_ptr<inorder_queue> q)
+: _q{std::move(q)} {}
```

Comment:
Don't we need

```suggestion
: _q{std::move(q)}, _num_submitted_operations{0} {}
```

Reply:
Absolutely, we have to initialize it to 0. But I don't think this can impact multi-gpu cases. Does it have an effect for your issue?

Reply:
No impact, but I noticed it while looking at the code.